Docs: add headless mode and Solana batch guides (#2867)
MetaMask/metamask-docs: metamask-connect/multichain/guides/headless-mode.md

---
title: 'Use Headless Mode - MetaMask Connect Multichain'
sidebar_label: Use headless mode
description: Render a custom QR code or connection UI by using headless mode with MetaMask Connect Multichain.
keywords:
  [
    headless,
    QR code,
    custom UI,
    display_uri,
    multichain,
    MetaMask,
    Connect,
    connection,
    modal,
    deeplink,
  ]
---

> # Use headless mode

Reviewer: Is this also supported on the EVM client?

Author: Yes, it's supported, EVM, yes, let me confirm for Solana. Do you think we should create a separate headless mode guide for EVM, too? Have mentioned it briefly in the Migration guide from SDK to MM Connect.

Author: Not a blocker for this PR

By default, MetaMask Connect renders its own QR code modal when connecting to MetaMask Mobile
via MetaMask Wallet Protocol (MWP).
Headless mode suppresses this built-in modal so you can render your own connection UI.

Use headless mode when you want to:

- Display a custom-styled QR code that matches your dapp's design.
- Show the connection URI in a different format (for example, a deeplink button instead of a QR code).
- Integrate the connection flow into an existing modal or onboarding wizard.

## Set up headless mode

### 1. Create the client with headless mode enabled

> Set `ui.headless` to `true` in the client options:

Reviewer: ```suggestion
### 1. Configure the client with headless mode

Initialize a MetaMask Connect Multichain client, and set `ui.headless` to `true` in the configuration options:
```

> Set `ui.headless` to `true` in the client options:

Reviewer: Should we add back reference documentation for MM Connect options? Do these overlap across the clients?

Author: Just for EVM and Multichain, will confirm for Solana

```javascript
import { createMultichainClient, getInfuraRpcUrls } from '@metamask/connect-multichain'

const client = await createMultichainClient({
  dapp: {

>     name: 'My DApp',

Reviewer: ```suggestion
    name: 'My Dapp',
```

    url: window.location.href,
  },
  api: {
    supportedNetworks: {
      ...getInfuraRpcUrls({ infuraApiKey: '<YOUR_INFURA_API_KEY>' }),
    },
  },
  ui: { headless: true },
})
```

### 2. Register a `display_uri` listener before connecting

The `display_uri` event fires during the connecting phase with a one-time-use pairing URI.

> You **must** register the listener before calling `connect()`, or you may miss the event:

Reviewer: ```suggestion
You **must** register the listener before calling `connect`, or you may miss the event:
```

We recently added a style rule to omit empty parentheses when referring to methods in prose.

```javascript
client.on('display_uri', uri => {
  showCustomQrModal(uri)
})
```

### 3. Connect and handle the result

```javascript
try {
  await client.connect(['eip155:1', 'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp'], [])
  hideCustomQrModal()
} catch (err) {
  hideCustomQrModal()
  if (err.code === 4001) {
    // User rejected — show retry UI
  } else {
    console.error('Connection failed:', err)
  }
}
```

## Important considerations

### URI is one-time-use

The pairing URI delivered by `display_uri` is a one-time-use token.
Once used or expired, it cannot be reused.

> If the connection fails, start a new `connect()` call to generate a fresh URI.

Reviewer: ```suggestion
If the connection fails, call `connect` again to generate a fresh URI.
```

### `display_uri` only fires during connecting

The event fires only while the client status is `'connecting'`.
After the connection resolves (success or error), `display_uri` stops firing.

### Extension connections skip QR

When the MetaMask browser extension is installed and `ui.preferExtension` is `true` (the default),
the SDK connects directly through the extension.
No `display_uri` event fires because no QR code is needed.

To force the MWP/QR flow even when the extension is available, set `ui.preferExtension` to `false`:

```javascript
const client = await createMultichainClient({

>   dapp: { name: 'My DApp', url: window.location.href },

Reviewer: ```suggestion
To display the QR connection option even when the extension is available, set `ui.preferExtension` to `false`:

```javascript
const client = await createMultichainClient({
  dapp: { name: 'My Dapp', url: window.location.href },
```

I assume the QR connection is displayed as an option, not as the required connection method? i.e. "force" sounds too strong.

Author: But this is one edge case, if someone wants to show QR code, even if the MM Wallet extension is available, they have to do this step, force would be the right word here.

Author: MWP is MetaMask Web Protocol, and we need some education on this to help with debugging. Let me attempt a rewrite for this part.

  ui: {
    headless: true,
    preferExtension: false,
  },
})
```

### Monitor connection status

> Use the `stateChanged` event to track the connection lifecycle and update your UI accordingly:

Reviewer: Should `stateChanged` be documented in the provider API reference (or elsewhere)?

Author: It's documented [here](https://metamask-docs-idf9vhkz7-consensys-ddffed67.vercel.app/metamask-connect/multichain/reference/methods/#events) under Events of MM Connect Multichain methods

Author: hyperlinked `stateChanged` to right place

[d80a0ad](https://github.com/MetaMask/metamask-docs/pull/2867/commits/d80a0ad998d163ce14c02c4340a055d40b65d8a8)
